File: backend/app/schemas/catalog.py
from datetime import datetime
from enum import Enum
from typing import Generic, TypeVar

from pydantic import BaseModel

T = TypeVar("T")


class AuthorOut(BaseModel):
    id: str
    name: str
    slug: str


class CategoryRef(BaseModel):
    id: str
    name: str
    slug: str


class CategoryOut(CategoryRef):
    description: str
    book_count: int


class BookSummary(BaseModel):
    """Lo necesario para una tarjeta de libro. No incluye la descripción larga."""

    id: str
    slug: str
    title: str
    author: AuthorOut
    categories: list[CategoryRef]

    price_cop: int  # precio de lista
    discount_pct: int
    final_price_cop: int  # precio a pagar por unidad, ya con descuento
    shipping_cost_cop: int
    free_shipping: bool

    stock: int
    in_stock: bool

    cover: str
    cover_width: int
    cover_height: int
    cover_blur: str | None

    featured: bool
    bestseller: bool
    rating_avg: float | None  # None si aún no hay reseñas
    rating_count: int


class BookDetail(BookSummary):
    publisher: str | None
    pages: int | None
    language: str | None
    description: str
    incomplete: bool


class ReviewOut(BaseModel):
    id: str
    author_name: str
    rating: int
    comment: str
    is_demo: bool
    verified: bool  # reseña de una cuenta que compró el libro (las de muestra nunca lo son)
    created_at: datetime
    updated_at: datetime | None = None


class Page(BaseModel, Generic[T]):
    items: list[T]
    total: int
    page: int
    page_size: int
    pages: int


class BookSort(str, Enum):
    relevance = "relevance"
    price_asc = "price_asc"
    price_desc = "price_desc"
    rating = "rating"
    title = "title"
    newest = "newest"

File: frontend/src/types/catalog.ts
// Espejo de backend/app/schemas/catalog.py. Si cambia allí, cambia aquí.

export type Author = { id: string; name: string; slug: string };

export type CategoryRef = { id: string; name: string; slug: string };

export type Category = CategoryRef & { description: string; book_count: number };

export type BookSummary = {
  id: string;
  slug: string;
  title: string;
  author: Author;
  categories: CategoryRef[];
  price_cop: number; // precio de lista
  discount_pct: number;
  final_price_cop: number; // por unidad, con descuento
  shipping_cost_cop: number;
  free_shipping: boolean;
  stock: number;
  in_stock: boolean;
  cover: string;
  cover_width: number;
  cover_height: number;
  cover_blur: string | null;
  featured: boolean;
  bestseller: boolean;
  rating_avg: number | null;
  rating_count: number;
};

export type BookDetail = BookSummary & {
  publisher: string | null;
  pages: number | null;
  language: string | null;
  description: string;
  incomplete: boolean;
};

export type Review = {
  id: string;
  author_name: string;
  rating: number;
  comment: string;
  is_demo: boolean;
  verified: boolean; // de una cuenta que compró el libro
  created_at: string;
  updated_at: string | null;
};

/** Situación de la persona con sesión frente a un libro (GET /books/{slug}/reviews/mine). */
export type MyReview = { can_review: boolean; has_purchased: boolean; review: Review | null };

export type Page<T> = {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
};

export type BookSort = "relevance" | "price_asc" | "price_desc" | "rating" | "title" | "newest";
